ENH: add type stubs from numpy-stubs

person142 opened the pull request:
Add the type stubs and tests from numpy-stubs. See the mailing list discussion about doing this here:

http://numpy-discussion.10968.n7.nabble.com/Put-type-annotations-in-NumPy-proper-tp47996.html

and also the discussion here:

https://github.com/numpy/numpy-stubs/issues/79

Things this entails:

- Copy over the stubs (`numpy/__init__.pyi`, `numpy/typing.pyi`, and `numpy/core/_internal.pyi`)
  - The only modification made was removing `ndarray.tostring` since it is deprecated
- Update some `setup.py` files to include pyi files
- Move the tests from `numpy-stubs/tests` into `numpy/tests`
  - Skip them if mypy is not installed (planning on setting up CI in a
    future PR)
  - Add a `mypy.ini`; use it to configure mypy in the tests
    - It tells mypy where to find NumPy in the test env
    - It ignores internal NumPy type errors (since we only want to consider errors from the tests cases)
  - Some small edits were made to fix test cases that were emitting deprecation warnings
 - Add `numpy/py.typed` so that the types are picked up in an installed version of NumPy

Still to do in follow up PRs:

- ~Install mypy in one of the CI runs so that the typing tests are run at a regular cadence~
- Make the `typing` module exist at runtime; see e.g. the discussion here:

  http://numpy-discussion.10968.n7.nabble.com/Feelings-about-type-aliases-in-NumPy-tp48059.html

  and here:

  https://github.com/numpy/numpy-stubs/pull/66

person142 committed:
ENH: add type stubs from numpy-stubs

Add the type stubs and tests from numpy-stubs. Things this entails:

- Copy over the stubs (numpy/__init__.pyi and
  numpy/core/_internal.pyi)
  - The only modification made was removing `ndarray.tostring` since
    it is deprecated
- Update some setup.py files to include pyi files
- Move the tests from numpy-stubs/tests into numpy/tests
  - Skip them if mypy is not installed (planning on setting up CI in a
    future PR)
  - Add a mypy.ini; use it to configure mypy in the tests
    - It tells mypy where to find NumPy in the test env
    - It ignores internal NumPy type errors (since we only want to
      consider errors from the tests cases)
  - Some small edits were made to fix test cases that were emitting
    deprecation warnings
 - Add numpy/py.typed so that the types are picked up in an
   installed version of NumPy

person142 commented on numpy/__init__.pyi:
```diff
@@ -0,0 +1,1077 @@
+import builtins
```
Big copy-paste from https://github.com/numpy/numpy-stubs/blob/master/numpy-stubs/__init__.pyi.

person142 commented on numpy/core/_internal.pyi:
```diff
@@ -0,0 +1,18 @@
+from typing import Any
```
Copy-paste from https://github.com/numpy/numpy-stubs/blob/master/numpy-stubs/core/_internal.pyi.

person142 commented on numpy/tests/fail/array_like.py:
```diff
@@ -0,0 +1,22 @@
+from typing import Any, TYPE_CHECKING
```
Tests are a copy paste from https://github.com/numpy/numpy-stubs/tree/master/tests.

person142 commented on numpy/tests/test_typing.py:
```diff
@@ -0,0 +1,127 @@
+import importlib.util
```
Test runner is a copy paste from https://github.com/numpy/numpy-stubs/blob/master/tests/test_stubs.py, but with some modifications:

- In the `numpy-stubs` repo the tests were run against an installed version of the stubs (because you can't point at `*-stubs` repo using `MYPYPATH`)
- Now we can use `MYPYPATH` to point at NumPy in the test env, so do that so that the tests can run in the normal way
- This does mean that mypy is type checking all of NumPy, so we will get complaints about NumPy internals that we don't want to hear
- To fix that we use a `mypy.ini` file that silences them
- By default errors from inside installed third-party libraries are silenced, so those internal type errors shouldn't bother end-users.

person142 commented on numpy/tests/test_typing.py:
```diff
@@ -0,0 +1,127 @@
+                )
+
+
+@pytest.mark.skipif(NO_MYPY, reason="Mypy is not installed")
```
Skip if mypy isn't installed, since we want to keep typing purely opt-in.

person142 commented on numpy/typing.pyi:
```diff
@@ -0,0 +1,64 @@
+import sys
```
Copy paste of https://github.com/numpy/numpy-stubs/blob/master/numpy-stubs/typing.pyi.

Next step is to make this available at runtime too.

person142 commented on typing_requirements.txt:
```diff
@@ -0,0 +1,2 @@
+mypy==0.770
```
If you install these requirements, then the typing tests should run and pass for you locally (I hope).

mattip replied:
mypy uses parts of the C-API that PyPy has not implemented, so this should be (not tested,

```suggestion
mypy==0.770; platform_python_implementation != "PyPy"
```

mattip reviewed (commented):
This is nice, thanks. I am not sure about the test directory structure, since it layers the tests into the already-existing `numpy/tests/*`. Maybe they should be in a separate directory like `numpy/tests/typing/*` and only the `numpy/tests/test_typing.py` left in the upper-level directory?

I think we should add this to the regular CI runs. If it takes too long we could mark it whit `@slow`. All we need to do then is merge the requirements to the current `test_requirements.txt` file.

I assume this does not slow down `import numpy` at all, is that correct?

person142 committed:
MAINT: move typing tests cases into a subdirectory of numpy/tests

person142 committed:
MAINT: merge typing requirements into test requirements

This will make them get picked up in CI runs.

rgommers reviewed (approved):
This approach LGTM, +1 to merge once CI is green.

person142 commented:
> Maybe they should be in a separate directory like numpy/tests/typing/* and only the numpy/tests/test_typing.py left in the upper-level directory?

Done in https://github.com/numpy/numpy/pull/16515/commits/3fee17a0da9277ee21e20018beeb2a42ef12a02f. Moved the `mypy.ini` into that dir too.

> All we need to do then is merge the requirements to the current test_requirements.txt file.

Done in https://github.com/numpy/numpy/pull/16515/commits/744512dc5d8c55b4873c3dd71a82b7915a886965.

> If it takes too long we could mark it whit @slow

After the cache is warm they aren't _too_ bad, eventually we can pass the cache along as a CI artifact. They also currently invoke mypy once per file, which slows things down (in the stubs this wasn't an issue because there was much less code for mypy to look at). We can speed that up by running mypy once over the entire `tests/typing` directory in say a pytest module-scoped fixture and then farming the results out to the various checks. tl;dr I guess is they are a little slow right now, but we've got a lot of room for tricks to speed them up.

> I assume this does not slow down import numpy at all, is that correct?

Correct. If you had bunch of annotations not in stubs files then I think it could, though I think that won't be a problem in 3.8 with `from __future__ import annotations`.

charris commented:
Needs a release note.

person142 committed:
MAINT: modify mypy error line parsing to work on windows

It currently splits on ":", which causes problems with drives.

person142 committed:
DOC: Add a release note about types being added

person142 commented:
> Needs a release note.

Added in https://github.com/numpy/numpy/pull/16515/commits/7c0d99d99c2034930e558d8b0fb48869964a41f0.

person142 commented:
Ok, getting failures on Python 3.9, but it looks like mypy itself is failing on 3.9 (at least as of writing this):

https://travis-ci.org/github/python/mypy

so I'm probably going to skip these tests on 3.9 for now.

person142 committed:
MAINT: avoid divide by zero in typing test causing aarch64 tests to fail

person142 committed:
MAINT: only run typing tests on python > 3.6 and < 3.9

Mypy doesn't work with 3.9 yet, and 3.6 doesn't work because it
doesn't the py.typed marker.

person142 commented:
Ah, and Python 3.6 doesn't work because `py.typed` wasn't a thing yet (if you look at the CI failures it is just completely failing to find any type information on the 3.6 run). I've restricted to testing on 3.7 and 3.8.

person142 committed:
MAINT: avoid more division by zero in typing test causing aarch64 to fail

mattip commented on numpy/tests/typing/pass/simple.py:
```diff
@@ -0,0 +1,175 @@
+array *= 1
+
+array / 1
+1 / array
```
This line is still failing on arm64

mattip commented:
The travis failure was fixed on master, so restarting this with just the fix for arm64 should get further: no rebase needed.

person142 commented:
Ok, CI is all green!

eric-wieser commented:
What's the plan for keeping this in sync with numpy-stubs? Is numpy-stubs going to be retired?

person142 commented:
> What's the plan for keeping this in sync with numpy-stubs? Is numpy-stubs going to be retired?

Yes, post-merge I'm planning on updating the numpy-stubs README to point users to the main repo and then archiving numpy-stubs.

BvB93 commented:
> Yes, post-merge I'm planning on updating the numpy-stubs README to point users to the main repo and then archiving numpy-stubs.

If that's the case, might it be an idea to add a dedicated typing label to the numpy repo?

mattip commented:
@BvB93 what is a "dedicated typing label"?

person142 commented:
I think what is meant is a Github label along the lines of the one @tylerjereddy made for SciPy:

https://github.com/scipy/scipy/pulls?q=label%3A%22static+typing%22+

mattip commented:
Thanks @person142. Looking forward to seeing how this works

anirudh2290 commented on numpy/typing.pyi:
```diff
@@ -0,0 +1,64 @@
+    @overload
+    def __array__(self, dtype: DtypeLike = ...) -> ndarray: ...
+
+ArrayLike = Union[bool, int, float, complex, _SupportsArray, Sequence]
```
do we support `__array_interface__`  interface for ArrayLike ?
